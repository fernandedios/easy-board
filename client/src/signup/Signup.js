import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';

import FormValidator from '../utils/FormValidator';
import Button from '../common/Button';
import TextInput from '../common/TextInput';

class Signup extends Component {
    constructor(props) {
        super(props);

        this.submitted = false;
        this.validator = new FormValidator([
            {
                field: 'name',
                method: 'isEmpty',
                validWhen: false,
                message: 'Name is required.'
            },
            {
                field: 'email',
                method: 'isEmpty',
                validWhen: false,
                message: 'Email address is required.'
            },
            {
                field: 'email',
                method: 'isEmail',
                validWhen: true,
                message: 'Please provide a valid email address'
            },
            {
                field: 'password',
                method: 'isEmpty',
                validWhen: false,
                message: 'Password is required'
            },
            {
                field: 'password',
                method: this.passwordValid,
                validWhen: true,
                message: 'Minimum 6 characters, one number, one lowercase and one uppercase letter'
            },
            {
                field: 'confirm_password',
                method: 'isEmpty',
                validWhen: false,
                message: 'Confirm Password is required'
            },
            {
                field: 'confirm_password',
                method: this.passwordMatch,
                validWhen: true,
                message: 'Passwords do not match'
            }

        ]);

        this.state = {
            name: '',
            email: '',
            password: '',
            confirm_password: '',
            avatar: '',
            role: '',
            validation: this.validator.valid()
        };

        this.handleEvent = this.handleEvent.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    passwordMatch = (confirmation, state) => (state.password === confirmation)
    passwordValid(state) {
        // at least one number, one lowercase and one uppercase letter
        // at least six characters
        const re = /(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,}/;
        return re.test(state.password);
    } 

    handleEvent(e) {
        e.preventDefault();
        this.setState({ [e.target.name]: e.target.value });
    }

    onSubmit(e) {
        e.preventDefault();

        const validation = this.validator.validate(this.state);
        this.setState({ validation });
        this.submitted = true;

        if (validation.isValid) {
            const { name, email, avatar, password } = this.state;
            const query = `mutation {
                signup (name: ${name}, email: ${email}, password: ${password}, avatar: ${avatar}, role: "user")
            }`;
            axios.post(`${process.env.REACT_APP_API}/api`, { query })
                .then(response => {
                    console.log(response);
                })
                .catch(error => {
                    console.log(error);
                });
        }
    }

    render() {
        let validation = this.submitted ? this.validator.validate(this.state) : this.state.validation;

        return (
            <section id="signup-box" className="form-box hero">
                <div className="hero-body">
                    <div className="container has-text-centered">
                        <div className="column is-6 is-offset-3">
                            <h3>Signup to Easy Board</h3>
                            <p>Create an account to join</p>
                            <div className="box">
                                <form onSubmit={this.onSubmit}>
                                    <TextInput message={validation.name.message} onChange={this.handleEvent} value={this.state.name} name="name" type="text" placeholder="Your Name" />
                                    <TextInput message={validation.email.message} onChange={this.handleEvent} value={this.state.email} name="email" type="email" placeholder="Email Address" />
                                    <TextInput onChange={this.handleEvent} value={this.state.avatar} name="avatar" type="text" placeholder="Profile Avatar" />
                                    <TextInput message={validation.password.message} onChange={this.handleEvent} value={this.state.password} name="password" type="password" placeholder="Password" />
                                    <TextInput message={validation.confirm_password.message} onChange={this.handleEvent} value={this.state.confirm_password} name="confirm_password" type="password" placeholder="Confirm Password" />
                                    <Button type="submit" name="signup-btn" classes="is-block is-info is-fullwidth" label="Signup" />
                                </form>
                            </div>
                            <p className="has-text-grey">Already have an account? <Link to="/login">Login here</Link></p>
                        </div>
                    </div>
                </div>
            </section>
        );
    }
}

export default Signup;
