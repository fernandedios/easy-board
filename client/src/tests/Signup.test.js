import React from 'react';
import { shallow, mount } from 'enzyme';
import ReactTestRenderer from 'react-test-renderer';

import Signup from '../signup/Signup';

describe('<SignupForm />', () => {
    it('should be defined', () => {
        expect(Signup).toBeDefined();
    });

    it('should have correct classNames', () => {
        const output = shallow(<Signup />);
        expect(output.find('section').hasClass('form-box')).toEqual(true);
        expect(output.find('section').hasClass('hero')).toEqual(true);
    });

    it('should render input fields correctly', () => {
        const output = mount(<Signup />);

        expect(output.find('#name').length).toEqual(1);
        expect(output.find('#email').length).toEqual(1);
        expect(output.find('#avatar').length).toEqual(1);
        expect(output.find('#password').length).toEqual(1);
        expect(output.find('#confirm_password').length).toEqual(1);
    });

    it('should render correctly', () => {
        const output = ReactTestRenderer.create(<Signup />).toJSON();
        expect(output).toMatchSnapshot();
    });
});

describe('Name Input Field', () => {
    it('should be a controlled component that responds to state change', () => {
        const output = mount(<Signup />);
        const name = 'Fernan';

        output.find('#name').simulate('change', {
            target: { name: 'name', value: name }
        });

        expect(output.state('name')).toEqual(name);
    });
});

describe('Email Input Field', () => {
    it('should be a controlled component that responds to state change', () => {
        const output = mount(<Signup />);
        const email = 'fernan@test.com';

        output.find('#email').simulate('change', {
            target: { name: 'email', value: email }
        });

        expect(output.state('email')).toEqual(email);
    });
});

describe('Password Input Field', () => {
    it('should be a controlled component that responds to state change', () => {
        const output = mount(<Signup />);
        const pass = 'pass123';

        output.find('#password').simulate('change', {
            target: { name: 'password', value: pass }
        });

        expect(output.state('password')).toEqual(pass);
    });
});

describe('Confirm Password Input Field', () => {
    it('should be a controlled component that responds to state change', () => {
        const output = mount(<Signup />);
        const pass = 'pass123';

        output.find('#confirm_password').simulate('change', {
            target: { name: 'confirm_password', value: pass }
        });

        expect(output.state('confirm_password')).toEqual(pass);
    });
});
