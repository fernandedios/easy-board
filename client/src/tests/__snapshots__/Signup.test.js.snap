// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupForm /> should render correctly 1`] = `
<section
  className="form-box hero"
  id="signup-box"
>
  <div
    className="hero-body"
  >
    <div
      className="container has-text-centered"
    >
      <div
        className="column is-6 is-offset-3"
      >
        <h3>
          Signup to Easy Board
        </h3>
        <p>
          Create an account to join
        </p>
        <div
          className="box"
        >
          <form
            onSubmit={[Function]}
          >
            <div
              className="field"
            >
              <label
                className="label"
              />
              <div
                className="control has-icons-right"
              >
                <input
                  className="input"
                  id="name"
                  name="name"
                  onChange={[Function]}
                  placeholder="Your Name"
                  type="text"
                  value=""
                />
              </div>
              <p
                className="help is-danger"
              >
                
              </p>
            </div>
            <div
              className="field"
            >
              <label
                className="label"
              />
              <div
                className="control has-icons-right"
              >
                <input
                  className="input"
                  id="email"
                  name="email"
                  onChange={[Function]}
                  placeholder="Email Address"
                  type="email"
                  value=""
                />
              </div>
              <p
                className="help is-danger"
              >
                
              </p>
            </div>
            <div
              className="field"
            >
              <label
                className="label"
              />
              <div
                className="control has-icons-right"
              >
                <input
                  className="input"
                  id="avatar"
                  name="avatar"
                  onChange={[Function]}
                  placeholder="Profile Avatar"
                  type="text"
                  value=""
                />
              </div>
              <p
                className="help is-danger"
              >
                
              </p>
            </div>
            <div
              className="field"
            >
              <label
                className="label"
              />
              <div
                className="control has-icons-right"
              >
                <input
                  className="input"
                  id="password"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
              <p
                className="help is-danger"
              >
                
              </p>
            </div>
            <div
              className="field"
            >
              <label
                className="label"
              />
              <div
                className="control has-icons-right"
              >
                <input
                  className="input"
                  id="confirm_password"
                  name="confirm_password"
                  onChange={[Function]}
                  placeholder="Confirm Password"
                  type="password"
                  value=""
                />
              </div>
              <p
                className="help is-danger"
              >
                
              </p>
            </div>
            <button
              className="button is-block is-info is-fullwidth"
              id="signup-btn"
              name="signup-btn"
              type="submit"
            >
              Login
            </button>
          </form>
        </div>
        <p
          className="has-text-grey"
        >
          Already have account? 
          <a
            href="/login"
          >
            Login here
          </a>
        </p>
      </div>
    </div>
  </div>
</section>
`;
